### PR TITLE
Tensorflow Writing Algorithm

### PR DESCRIPTION
**Pull request by LouisSzeto**
#### Description
Add tutorial for `Tensorflow` library on [Writing Algorithm/Machine Learning](https://github.com/QuantConnect/Documentation/tree/master/02%20Writing%20Algorithms/30%20Machine%20Learning/03%20Popular%20Libraries).

#### Motivation and Context
Easy implementation example for users.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [X] New feature (non-breaking change which adds functionality)
- [X] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [X] My code follows the code style of this project.
- [X] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [X] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/01 Introduction.html [hunk 1]
```diff
@@ -0,0 +1 @@
+<p>This page explains how to build, train, deploy and store <code>Tensorflow</code> v1 models. LEAN now supports <code>Tensorflow 2</code>, refer to <a href="">Keras</a> tutorial deploy that.</p>
```
Comment (DerekMelchin) [inline 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/01 Introduction.html:1]: Change the second sentence to 

> To view the tutorial on Tensorflow 2, see [Keras](https://www.quantconnect.com/docs/v2/writing-algorithms/machine-learning/popular-libraries/keras).

File: 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/03 Create Subscriptions.html [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<h4>Create Subscriptions</h4>
```
Comment (DerekMelchin) [inline 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/03 Create Subscriptions.html:1]: Remove this h4

File: 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/03 Create Subscriptions.html [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+<p>In the <code>Initialize</code> method, <a href="https://www.quantconnect.com/docs/v2/writing-algorithms/securities/asset-classes">subscribe to some data</a> so you can train the <code>tensorflow</code> model and make predictions.</p>
+<p>You need to <a href="/docs/v2/writing-algorithms/securities/asset-classes">subscribe to some data</a> to train the model.</p>
```
Comment (DerekMelchin) [inline 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/03 Create Subscriptions.html:3]: Delete this line, it's a duplicate

File: 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/04 Build Models.html [hunk 4]
```diff
@@ -0,0 +1,81 @@
+<p>In this example, a neural-network regression prediction model using the following features and labels:</p>
```
Comment (DerekMelchin) [inline 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/04 Build Models.html:1]: > In this example, build a neural-network regression prediction model that uses the following features and labels:

File: 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/04 Build Models.html [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    bias_hidden_3 = tf.Variable(bias_initializer([num_neurons_3]))
+    hidden_3 = tf.nn.relu(tf.add(tf.matmul(hidden_2, W_hidden_3), bias_hidden_3))
+    
+    # Create the output layer and give it a name, so it is accessible after saving and loading the model.
+    W_out = tf.Variable(weight_initializer([num_neurons_3, 1]))
+    bias_out = tf.Variable(bias_initializer([1]))
+    output = tf.transpose(tf.add(tf.matmul(hidden_3, W_out), bias_out), name='outer')
+    
+    # Set up the loss function and optimizers for gradient descent optimization and backpropagation.
+    # This example uses mean-square error as the loss function because the close price is a continuous data and use Adam as the optimizer because of its adaptive step size.
```
Comment (DerekMelchin) [inline 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/04 Build Models.html:65]: "use" => "uses"

File: 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/04 Build Models.html [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    loss = tf.reduce_mean(tf.squared_difference(output, Y))
+    optimizer = tf.train.AdamOptimizer().minimize(loss)
+    
+    return sess, X, Y, output, optimizer</pre>
+    </div>
+    
+    <li>Instantiate the model, input layers, output layer, and optimizer. Save them as class variables.</li>
```
Comment (DerekMelchin) [inline 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/04 Build Models.html:72]: ". Save" => "and then save"

File: 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/06 Predict Labels.html [hunk 7]
```diff
@@ -0,0 +1,14 @@
+<p>To predict the labels of new data, in the <code>OnData</code> method, get the most recent set of features and then call the <code>run</code> method with new features.</p>
+<div class="section-example-container">
+    <pre class="python">new_features, __ = self.get_features_and_labels()
+prediction = self.model.run(self.output, feed_dict={self.X: new_features})
```
Comment (DerekMelchin) [inline 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/06 Predict Labels.html:4]: Instead of making predictions on all the feature samples, just use the last row (the most recent 5 closing prices)

File: 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/06 Predict Labels.html [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+prediction = float(prediction.flatten()[-1])</pre>
+</div>
+
+<p>You can use the label prediction to place orders.</p>
+<div class="section-example-container">
+    <pre class="python">if prediction > 0:
```
Comment (DerekMelchin) [inline 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/06 Predict Labels.html:10]: If the label (`prediction`) is the next closing price, shouldn't we check if it's greater than the current price, not if it's greater than 0?

File: 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/07 Save Models.html [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+<ol>
+    <li>Export the <code>TensorFlow</code> graph as a JSON object.</li>
+    <div class="section-example-container">
+        <pre class="python">graph_definition = tf.compat.v1.train.export_meta_graph()
+json_graph = json_format.MessageToJson(graph_definition)</pre>
+    </div>
+
+    <li>Export the <code>TensorFlow</code> weights as a JSON object.</li>
+    <div class="section-example-container">
+        <pre class="python"># Define a function to get the weights from the tensorflow session
```
Comment (DerekMelchin) [inline 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/07 Save Models.html:12]: I think we should remove this internal function definition. If we just use the body of the definition, we can reduce the logic down to just 3 lines of code.

```
    weights = self.model.run(tf.compat.v1.trainable_variables())
    weights = [w.tolist() for w in weights]
    json_weights = json5.dumps(weights)
```

File: 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/08 Load Models.html [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    value = np.asarray(value)
+    assign_placeholder = tf.placeholder(var.dtype, shape=value.shape)
+    assign_op = var.assign(assign_placeholder)
+    assign_ops.append(assign_op)
+    feed_dict[assign_placeholder] = value
+self.model.run(assign_ops, feed_dict=feed_dict)</pre>
+    </div>
+</ol>
+
+<p>You can load and trade with pre-trained <code>tensorflow</code> models that you saved in the ObjectStore. To load a <code>tensorflow</code> model from the ObjectStore, in the <code>Initialize</code> method, get the file path to the saved model and then recall the graph and weights of the model.</p>
```
Comment (DerekMelchin) [inline 02 Writing Algorithms/30 Machine Learning/03 Popular Libraries/02 Tensorflow/08 Load Models.html:47]: Delete everything above this line, duplicate content.

**Review by DerekMelchin — APPROVED** [review]
